python/semantic_kernel/connectors/search/google/google_search.py

```diff
@@ -0,0 +1,194 @@
+            None
+            if not options.include_total_count
+            else response.search_information.total_results or None
+            if response.search_information
+            else None
+        )
+        if total_results is not None:
+            return int(total_results)
+        return None
+
+    def _get_options(self, options: "SearchOptions | None", **kwargs: Any) -> TextSearchOptions:
+        if options is not None and isinstance(options, TextSearchOptions):
+            return options
+        try:
+            return TextSearchOptions(**kwargs)
+        except ValidationError:
+            return TextSearchOptions()
+
+    async def _inner_search(self, query: str, options: TextSearchOptions) -> GoogleSearchResponse:
+        self._validate_options(options)
+
+        logger.info(
+            f"Received request for google web search with \
+                params:\nquery: {query}\nnum_results: {options.top}\noffset: {options.skip}"
+        )
+
+        full_url = f"{CUSTOM_SEARCH_URL}{self._build_query(query, options)}"
+        headers = {"user_agent": SEMANTIC_KERNEL_USER_AGENT}
+        try:
+            async with AsyncClient() as client:
+                response = await client.get(full_url, headers=headers)
+                response.raise_for_status()
+                return GoogleSearchResponse.model_validate_json(response.text)
+        except HTTPStatusError as ex:
+            logger.error(f"Failed to get search results: {ex}")
+            raise ServiceInvalidRequestError("Failed to get search results.") from ex
+        except RequestError as ex:
+            logger.error(f"Client error occurred: {ex}")
+            raise ServiceInvalidRequestError("A client error occurred while getting search results.") from ex
+        except Exception as ex:
+            logger.error(f"An unexpected error occurred: {ex}")
+            raise ServiceInvalidRequestError("An unexpected error occurred while getting search results.") from ex
+
+    def _validate_options(self, options: TextSearchOptions) -> None:
+        if options.top > 10:
+            raise ServiceInvalidRequestError("count value must be less than or equal to 10.")
+
+    def _build_query(self, query: str, options: TextSearchOptions) -> str:
+        params = {
+            "key": self.settings.api_key.get_secret_value(),
+            "cx": self.settings.engine_id,
+            "num": options.top,
+            "start": options.skip,
+        }
+        if options.filter:
+            for filter in options.filter.filters:
+                if isinstance(filter, EqualTo):
+                    if filter.field_name in QUERY_PARAMETERS:
+                        params[filter.field_name] = quote_plus(filter.value)
+                elif isinstance(filter, AnyTagsEqualTo):
+                    logger.debug("AnyTagEqualTo filter is not supported by Google Search API.")
+        return f"?q={quote_plus(query)}&{'&'.join(f'{k}={v}' for k, v in params.items())}"
```

Comment:
Is URL encoding of values needed?

Reply:
yes, this gets put straight into the URL not in body or params...